test/fs.js

```diff
@@ -0,0 +1,171 @@
+const path = require('path');
+const stream = require('stream');
+
+const testData =
+  'JySAF1figbONWxAeeoGhDXeI7df25GKXKPcyCl63ZjIv5HZGQH8yfFeHGGaSJ9zDMP0CwtgX2madIQkFmuq38SNhcKcGQ';
+const absoluteLink = __dirname.replace('test', 'lib');
+const srcFiles = [
+  'srcDir',
+  'srcDir/dir1',
+  'srcDir/dir2',
+  'srcDir/file1',
+  'srcDir/link',
+  'srcDir/dir1/subdir1',
+];
+
+const dstFiles = [
+  'dstDir',
+  'dstDir/dir1',
+  'dstDir/dir2',
+  'dstDir/file1',
+  'dstDir/link',
+  'dstDir/dir1/subdir1',
+];
+
+class ToUpperCase extends stream.Transform {
+  constructor(options = {}) {
+    options.decodeStrings = false;
+    super(options);
+  }
+  _transform(chunk, encoding, callback) {
+    if (encoding !== 'utf8') {
+      this.emit('error', new Error('only utf8'));
+      return callback();
+    }
+    this.push(chunk.toUpperCase());
+    return callback();
+  }
+}
+
+const transOptions = {
+  Transform: ToUpperCase,
+  readOptions: { encoding: 'utf-8' },
+  writeOptions: {},
+};
+
+const rmdirp = (dir, cb) => {
+  fs.lstat(dir, (err, stats) => {
+    if (err) {
+      if (err.code === 'ENOENT') cb(null);
+      else cb(err);
+      return;
+    }
+    if (stats.isDirectory()) {
+      fs.readdir(dir, (err, files) => {
+        if (err) {
+          cb(err);
+          return;
+        }
+        files = files.map(file => path.join(dir, file));
+        metasync.each(files, rmdirp, err => {
+          if (err) cb(err);
+          else fs.rmdir(dir, cb);
+        });
+      });
+    } else {
+      fs.unlink(dir, cb);
+    }
+  });
+};
+
+const copydirTest = metatests.test('Recursive dir copy');
+copydirTest.endAfterSubtests();
+
+copydirTest.beforeEach((test, cb) => {
+  metasync.series(
+    srcFiles,
+    (file, cb) => {
+      if (path.basename(file).startsWith('file')) {
+        fs.appendFile(file, testData, cb);
+      } else if (file.includes('link')) {
+        fs.symlink(absoluteLink, file, cb);
+      } else {
+        fs.mkdir(file, cb);
+      }
+    },
+    test.cb(cb)
+  );
+});
+
+copydirTest.afterEach((test, cb) => {
+  rmdirp(dstFiles[0], err => {
+    if (err) test.error(err);
+    rmdirp(srcFiles[0], test.cb(cb));
+  });
+});
+
+copydirTest.test('copy', test => {
+  common.copyDir(srcFiles[0], dstFiles[0], err => {
+    if (err) {
+      test.error(err);
+      test.end();
+      return;
+    }
+    metasync.each(dstFiles, fs.access, test.cb(() => test.end()));
+  });
+});
+
+copydirTest.test('dereference', test => {
+  common.copyDir(srcFiles[0], dstFiles[0], { dereference: true }, err => {
+    if (err) {
+      test.error(err);
+      test.end();
+      return;
+    }
+    metasync.each(dstFiles, fs.access, test.cb(() => test.end()));
+  });
+});
+
+copydirTest.test('transform', test => {
+  common.copyDir(srcFiles[0], dstFiles[0], transOptions, err => {
+    if (err) {
+      test.error(err);
+      test.end();
+      return;
+    }
+    metasync.each(dstFiles, fs.access, test.cb(() => test.end()));
```

Comment:
There should be check if contents of copied files were transformed.

Reply:
hmmm, Can you suggest me something?
```fs.readFile``` than compare it with ```.toUperCase()```?

Reply:
```js
metasync.each(dstFiles,
  (file, cb) => {
    const fileName = path.basename(file);
    if (fileName.startsWith('file')) {
      fs.readFile(file, (err, data) => {
        test.strictSame(data, expectedData);
        cb(err);
      });
    } else if (fileName.startsWith('link')) {
      // fs.access or fs.readFile depending on whether this is a file or directory
    } else {
      fs.access(file, cb);
    }
  },
  test.cb(() => test.end())
);
```

Reply:
I should do the almost same fn in each test
So it will be better to move it in separate function?

Reply:
Tests for `transform`, `dereference`, `filter`, `depth less than expected` should have different check functions since copying is performed differently, therefore, there is no need to create a separate function for this.